test/validation/api/scheduler/scheduler.c

```diff
@@ -1780,6 +1786,228 @@ static void scheduler_test_ordered_lock(void)
+
+static void scheduler_test_sched_and_plain(odp_schedule_sync_t sync)
+{
+	thread_args_t *args;
+	test_globals_t *globals;
+	odp_queue_t sched_queue;
+	odp_queue_t plain_queue;
+	odp_pool_t pool;
+	odp_queue_param_t queue_param;
+	odp_pool_param_t pool_param;
+	odp_queue_capability_t queue_capa;
+	odp_schedule_capability_t sched_capa;
+	odp_shm_t shm;
+	odp_event_t ev;
+	uint32_t *buf_data;
+	uint32_t seq;
+	uint64_t wait = odp_schedule_wait_time(100 * ODP_TIME_MSEC_IN_NS);
+	uint32_t events_per_queue = BUFS_PER_QUEUE / 2;
+	uint32_t prev_seq;
+	int first;
+
+	CU_ASSERT_FATAL(!odp_schedule_capability(&sched_capa));
+	CU_ASSERT_FATAL(!odp_queue_capability(&queue_capa))
+
+	if (sync == ODP_SCHED_SYNC_ORDERED &&
+	    sched_capa.max_ordered_locks == 0) {
+		printf("\n  NO ORDERED LOCKS. scheduler_test_ordered_and_plain skipped.\n");
+		return;
+	}
+
+	/* Set up the scheduling environment */
+	shm = odp_shm_lookup(GLOBALS_SHM_NAME);
+	CU_ASSERT_FATAL(shm != ODP_SHM_INVALID);
+	globals = odp_shm_addr(shm);
+	CU_ASSERT_PTR_NOT_NULL_FATAL(globals);
+
+	shm = odp_shm_lookup(SHM_THR_ARGS_NAME);
+	CU_ASSERT_FATAL(shm != ODP_SHM_INVALID);
+	args = odp_shm_addr(shm);
+	CU_ASSERT_PTR_NOT_NULL_FATAL(args);
+	args->globals = globals;
+
+	/* Make sure all events fit to queues */
+	if (sched_capa.max_queue_size &&
+	    sched_capa.max_queue_size < events_per_queue)
+		events_per_queue = sched_capa.max_queue_size;
+	if (queue_capa.plain.max_size &&
+	    queue_capa.plain.max_size < events_per_queue)
+		events_per_queue = queue_capa.plain.max_size;
+
+	odp_queue_param_init(&queue_param);
+	queue_param.type = ODP_QUEUE_TYPE_SCHED;
+	queue_param.sched.sync = sync;
+	queue_param.size = events_per_queue;
+	if (sync == ODP_SCHED_SYNC_ORDERED)
+		queue_param.sched.lock_count = 1;
+
+	sched_queue = odp_queue_create(NULL, &queue_param);
+	CU_ASSERT_FATAL(sched_queue != ODP_QUEUE_INVALID);
+	globals->sched_and_plain_q.sched = sched_queue;
+
+	odp_queue_param_init(&queue_param);
+	queue_param.type = ODP_QUEUE_TYPE_PLAIN;
+	queue_param.size = events_per_queue;
+
+	plain_queue = odp_queue_create(NULL, &queue_param);
+	CU_ASSERT_FATAL(sched_queue != ODP_QUEUE_INVALID);
+	globals->sched_and_plain_q.plain = plain_queue;
+
+	odp_pool_param_init(&pool_param);
+	pool_param.buf.size  = 100;
+	pool_param.buf.num   = 2 * events_per_queue;
+	pool_param.type      = ODP_POOL_BUFFER;
+
+	pool = odp_pool_create("sched_to_plain_pool", &pool_param);
+	CU_ASSERT_FATAL(pool != ODP_POOL_INVALID);
+
+	/* Create and enq test events with sequential sequence numbers */
+	for (seq = 0; seq < events_per_queue; seq++) {
+		odp_buffer_t buf1, buf2;
+
+		buf1 = odp_buffer_alloc(pool);
+		if (buf1 == ODP_BUFFER_INVALID)
+			break;
+		buf2 = odp_buffer_alloc(pool);
+		if (buf2 == ODP_BUFFER_INVALID) {
+			odp_buffer_free(buf1);
+			break;
+		}
+		buf_data = odp_buffer_addr(buf1);
+		*buf_data = seq;
+		buf_data = odp_buffer_addr(buf2);
+		*buf_data = seq;
+
+		CU_ASSERT_FATAL(!odp_queue_enq(sched_queue,
```

Comment:
Maybe important comment to add here is that this test depends on all events being of the same flow in case of a multi-flow scheduler.

Reply:
The flow id is zero by default, so multi-flow scheduling has no effect here.

Reply:
I know it has but it might be not so obvious to a person just reading the unit test. If you think the implied 0 is enough then please feel free to discard my comment.

Reply:
Okay, we are currently not referencing to multi-flow aware scheduler anywhere else except scheduler_test_flow_aware(), so adding the comment felt a bit out of place. But you gotta start improvements from somewhere, so I'll add a comment here.